File: findFirstAndLast/searchRange.go
package main

import (
	"fmt"
)

func main() {
	nums := []int{1, 3}
	fmt.Println(searchRange(nums, 3))
}

func searchRange(nums []int, target int) []int {
	first := -1
	last := -1
	l, r := 0, len(nums)-1

	for l <= r {
		mid := (r-l)/2 + l
		if nums[mid] == target {
			first, last = mid, mid
			for first > 0 && nums[first-1] == target {
				first--
			}
			for last < len(nums)-1 && nums[last+1] == target {
				last++
			}
			break
		} else if target < nums[mid] {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}

	return []int{first, last}
}

File: findFirstAndLast/searchRange.ts
function searchRange(nums: number[], target: number): number[] {

	return [searchTarget(nums, target, false), searchTarget(nums, target, true)]

};

function searchTarget(nums: number[], target: number, forward: boolean = false): number {
	let answer = -1
	let left = 0
	let right = nums.length - 1
	while (left <= right) {
		const mid = ~~((right - left) / 2) + left

		if (nums[mid] === target) {
			answer = mid
			if (forward) {
				left = mid + 1
			} else {
				right = mid - 1
			}


		} else if (nums[mid] > target) {
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return answer
}